src/contrib/mlir/backend/pass/affine_lowerer.cpp

```diff
@@ -678,6 +715,207 @@ namespace
+        NGRAPH_CHECK(lhsTy, "Unexpected non-memref LHS type");
+        NGRAPH_CHECK(rhsTy, "Unexpected non-memref RHS type");
+        NGRAPH_CHECK(biasTy, "Unexpected non-memref bias type");
+
+        Type elemTy = resultTy.getElementType();
+        NGRAPH_CHECK(elemTy == lhsTy.getElementType() && elemTy == rhsTy.getElementType() &&
+                         elemTy == biasTy.getElementType(),
+                     "Types mismatch in GemmOp");
+
+        MemRefView vRes(result), vLhs(lhs), vRhs(rhs), vBias(bias);
+
+        NGRAPH_CHECK(vLhs.rank() == 2 && vRhs.rank() == 2 && vRes.rank() == 2 && vBias.rank() <= 2,
+                     "Gemm operation is only supported for 2D tensors");
+
+        auto attrPtr = static_cast<gemmAttrs*>(malloc(sizeof(gemmAttrs)));
+        attrPtr->transposeA = gemm.transA();
+        attrPtr->transposeB = gemm.transB();
+        attrPtr->alpha = gemm.alpha().convertToFloat();
+        attrPtr->beta = gemm.beta().convertToFloat();
+        attrPtr->m = lhsShape[0];
+        attrPtr->k = lhsShape[1];
+        attrPtr->n = rhsShape[1];
+        attrPtr->lda = lhsShape[1];
+        attrPtr->ldb = rhsShape[1];
+
+        if (gemm.transA())
+        {
+            attrPtr->m = lhsShape[1];
+            attrPtr->k = lhsShape[0];
+        }
+        if (gemm.transB())
+        {
+            attrPtr->n = rhsShape[0];
+        }
+        attrPtr->ldc = attrPtr->n;
+
+        int broadcastHint;
+        if (vBias.rank() == 0)
+        {
+            // Scalar
+            broadcastHint = 2;
+        }
+        else if (vBias.rank() == 2)
+        {
+            if (biasShape[0] == attrPtr->m && biasShape[1] == 1)
+            {
+                broadcastHint = 1;
+            }
+            else if (biasShape[0] == 1 && biasShape[1] == attrPtr->n)
+            {
+                broadcastHint = 0;
+            }
+            else
+            {
+                broadcastHint = -1;
+            }
+        }
+        else
+        {
+            if (biasShape[0] == attrPtr->m)
+            {
+                broadcastHint = 1;
+            }
+            else if (biasShape[0] == attrPtr->n)
+            {
+                broadcastHint = 0;
+            }
+        }
+        attrPtr->broadcastHint = broadcastHint;
+
+        auto int64Ty = rewriter.getIntegerType(64);
+        auto unrankedMemrefTy = UnrankedMemRefType::get(elemTy, 0);
+        auto callBackFunc = pass.getCallDecl("__mlir_callback_3_inputs",
+                                             {unrankedMemrefTy,
+                                              unrankedMemrefTy,
+                                              unrankedMemrefTy,
+                                              unrankedMemrefTy,
+                                              int64Ty,
+                                              int64Ty},
+                                             {},
+                                             rewriter);
+
+        auto lhsCast =
```

Comment:
Can you create a helper to cast N arguments instead and place them in a SmallVectorImpl ? This code will get repetitive.